module/zfs/zfs_vnops.c

```diff
@@ -3563,6 +3652,17 @@ zfs_rename(struct inode *sdip, char *snm, struct inode *tdip, char *tnm,
 	}
 
 	/*
+	 * If we are using project inheritance, we only allow renames
+	 * into our tree when the project IDs are the same; otherwise
+	 * tree quota mechanism would be circumvented.
```

Comment:
I think you're saying that today there is no "tree quota" in ZFS, but we might want to support it someday.  When/if we support "tree quota", we will need to enforce that if PROJINHERIT is set, all descendants have the same project.

To make that easier to determine, would we also enforce that if PROJINHERIT is set, all descendant directories also have PROJINHERIT set?  That way if we are renaming a directory, if the new parent dir has PROJINHERIT, the dir that we're renaming must have the same PROJID and also have PROJINHERIT set (to ensure that all its descendants also have that projid).